prometheus_juju_backup_all_exporter/metrics.py

```diff
@@ -5,41 +5,56 @@
 
 
 class MetricBase(ABC):
-    """Abstract base class for creating custom collectors.
+    """Abstract base class for creating custom metrics.
 
-    All collector classes should add this class as a mixin class, then subclass
+    All metric classes should add this class as a mixin class, then subclass
     from the appropriate metric family that fits your needs. For example, you
-    can have a custom collector that looks like this:
+    can have a custom metric that looks like this:
 
     ```
-    class CustomCollector(MetricBase, GaugeMetricFamily):
+    from prometheus_client.core import CounterMetricFamily, GaugeMetricFamily
+
+    class CustomCounter(MetricBase, CounterMetricFamily):
         def __init__(self, *args, **kwargs):
             super().__init__(*args, **kwargs)
 
-        def _fetch(self, *args, **kwargs):
-            labels = ["test_label", "test_label_2"]
-            value = 10.0
-            return {"labels": labels, "value": value}
+        def _process(self, data, old_sample={}):
+            # Get previous count value from old_sample and add new count value
+            # to the counter.
+            for d in data:
+                d["value"] += old_sample.get(tuple(d["labels"]), 0.0)
+            return data
     ```
     """
 
     @abstractmethod
-    def _fetch(self):
-        """User defined fetch method used by `self.collect()`."""
-        pass
+    def _process(self, data, old_sample):
+        """User defined process method used by `self.add_samples()`.
+
+        Args:
+            data: the data to be processed
+            old_sample: previous metric sample, useful when implementing counter.
+
+        Returns:
+            processed data: list of dictionary
+        """
+        pass  # pragma: no cover
 
-    def __new__(cls, *args, **kwargs):
-        """Add logging message when creating new metrics."""
-        obj = super().__new__(cls)
-        logger.debug("Created new metric '{}'.".format(args[0]))
-        return obj
+    @property
+    def previous_data(self):
+        """Return previous sample as a labels <-> value pair."""
+        return {tuple(sample.labels.values()): sample.value for sample in self.samples}
 
     def __hash__(self):
-        """Return the hash of the custom collector."""
-        return hash(self.name)
+        """Return the hash of the custom metric."""
+        return hash(self.name)  # pragma: no cover
 
-    def collect(self):
+    def add_samples(self, metric_data, old_sample={}):
```

Comment:
[no-blocking]I believe provide a type annotation or example of input format in doc-string will be more helpful.

Reply:
This is actually internal function, not to be used by end user. So instead I added example input to `_process` and `_fecth` function